### PR TITLE
DEV: circleci caching for deps, build, doc

### PR DESCRIPTION
**Pull request by tupui**
Enable caching in circleci.

When iterating on a branch, CI could build again from a previous run. This could save a lot of CI and will improve feedback return time. This strategy is used by other packages like sklearn or skimage.

**EDIT:**

Caching of the doc itself does not work.

* use Meson
* ccache for build caching
* Add `[skip circle]` `[circle skip]`
* Remove pdf images
* Remove LaTeX
* Modernise the CircleCI config

### CODE REVIEW

**Comment by tupui** [comment]
This is in draft mode as I might test other things like trying to tar/untar the artifact.

I will do several force push too to check that the caching is working as expected and reduces the CI time.

**Comment by tupui** [comment]
@larsoner if think you can help here. I tried to add caching on circleci, but seems like building the doc is still slow. Not sure if this is because the machine is slow or something else. Looking at the log, caching seems to be in effect as it writes fast. It's spending most of the `build doc` step `waiting for workers...`

I suppose we could also improve the setup of the vm too. Can't we use a SciPy image here?

**Review by larsoner — COMMENTED** [review]
> I suppose we could also improve the setup of the vm too. Can't we use a SciPy image here?

Not that I can see -- most of the time is spent compiling, and for each PR you want to build *that version* of SciPy so compiling is necessary.

The key I think will be getting this PR to the stage where this says something else:
```
building [html]: targets for 300 source files that are out of date
updating environment: [new config] 3609 added, 0 changed, 0 removed
```
This indicates that Sphinx is going to rebuild everything from scratch. Once this says something like a handful of source files out of date, and no "new config" forcing a full rebuild, it will be much faster.

File: .circleci/config.yml [hunk 1]
```diff
@@ -42,6 +42,12 @@ jobs:
       - run:
           <<: *debian
 
+      - restore_cache:
+          keys:
+              - packages-{{ .Branch }}
+              - build-{{ .Branch }}
+              - doc-{{ .Branch }}
```
[the remark this answers is not in the document]
Comment (larsoner) [reply .circleci/config.yml:49]: ... also I wouldn't restore the docs here, but rather in the docs part of the CircleCI build

**Comment by tupui** [comment]
What I thought for the compilation and doc was to cache the first outcome of the CI. Then updating the PR would be faster.

**Comment by larsoner** [comment]
Technically it should be a little bit faster (in addition to more conceptually clean IMO) to restore the doc cache only during the `build_docs` job -- it has to happen somewhere, and by putting it in the `build_scipy` job, you delay both the `build_docs` job (which uses it) and the `run_benchmarks` job (which doesn't), since both depend on the `build_scipy` job

File: .circleci/config.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (tupui) [reply .circleci/config.yml]: 😒... Thanks for that! I was really puzzled... I will move it to the correct place.

**Comment by tupui** [comment]
> Technically it should be a little bit faster (in addition to more conceptually clean IMO) to restore the doc cache only during the `build_docs` job -- it has to happen somewhere, and by putting it in the `build_scipy` job, you delay both the `build_docs` job (which uses it) and the `run_benchmarks` job (which doesn't), since both depend on the `build_scipy` job

Agreed I will move this. Thanks.

**Comment by tupui** [comment]
Sorry @larsoner but I don't know what else to try here. It does not seems to work as expected. I might be setting some paths wrong or using the script wrong (the last attempt is without the script).

File: .circleci/config.yml [hunk 2]
```diff
@@ -75,16 +92,28 @@ jobs:
...
           command: |
             . venv/bin/activate
             export SHELL=$(which bash)
-            python -u runtests.py -g -j2 --doc html-scipyorg --doc latex
+            python setup.py install
+            cd doc
+            make html-scipyorg
+            make latex
             make -C doc/build/latex all-pdf LATEXOPTS="-file-line-error -halt-on-error"
             cp -f doc/build/latex/scipy-ref.pdf doc/build/html-scipyorg/
```
Comment (larsoner) [inline .circleci/config.yml:110]: With your changes you are in the `doc` dir so you need to do either (my preference):
```suggestion
            make -C doc html-scipyorg
            make -C doc latex
            make -C doc/build/latex all-pdf LATEXOPTS="-file-line-error -halt-on-error"
            cp -f doc/build/latex/scipy-ref.pdf doc/build/html-scipyorg/
```
or
```
            cd doc
            make html-scipyorg
            make latex
            make -C build/latex all-pdf LATEXOPTS="-file-line-error -halt-on-error"
            cp -f build/latex/scipy-ref.pdf doc/build/html-scipyorg/
```

Comment (tupui) [reply .circleci/config.yml]: Thanks for your help! Let's give this a try then 😄

**Comment by tupui** [comment]
ok this is still not working for some reason. Really puzzling and I am certain it's a path thing as locally you can build again both doc and sources and it's fast. Also here we lost the parallel thing seemingly, so we are even slower 😅

**Comment by larsoner** [comment]
In the log:
```
updating environment: [extensions changed ('sphinx_panels')] 3609 added, 0 changed, 0 removed
```
At first I thought this was a bug with `sphinx_panels`. But using numpydoc's `tinybuild` test infrastructure I can't replicate the problem, after adding `sphinx_panels` to the `conf.py` I see no full rebuild on the second run:

<details>

```
larsoner@bunk:~/python/numpydoc/numpydoc/tests/tinybuild$ make html
sphinx-build -nWT --keep-going -b html -d _build/doctrees . _build/html
Running Sphinx v3.5.4
making output directory... done
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
[autosummary] generating autosummary for: index.rst
[autosummary] generating autosummary for: /home/larsoner/python/numpydoc/numpydoc/tests/tinybuild/generated/numpydoc_test_module.MyClass.rst, /home/larsoner/python/numpydoc/numpydoc/tests/tinybuild/generated/numpydoc_test_module.my_function.rst
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: [new config] 3 added, 0 changed, 0 removed
reading sources... [100%] index                                                                                                                                                     
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] index                                                                                                                                                      
generating indices... genindex py-modindex done
writing additional pages... search done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded.

The HTML pages are in _build/html.
larsoner@bunk:~/python/numpydoc/numpydoc/tests/tinybuild$ make html
sphinx-build -nWT --keep-going -b html -d _build/doctrees . _build/html
Running Sphinx v3.5.4
loading pickled environment... done
[autosummary] generating autosummary for: generated/numpydoc_test_module.MyClass.rst, generated/numpydoc_test_module.my_function.rst, index.rst
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 0 source files that are out of date
updating environment: 0 added, 0 changed, 0 removed
looking for now-outdated files... none found
no targets are out of date.
build succeeded.

The HTML pages are in _build/html.

```

</details>

My guess is that a cache was created on a previous run that didn't have sphinx_panels, and was restored in this PR. Now when you run it you get the old version.

Can you replicate the problem locally? If you do `make clean; make html; make html` does the second `make html` report that it's rewriting all pages? This will tell us if it's a CircleCI stale cache problem or a scipy<->sphinx-pannels problem.

**Comment by larsoner** [comment]
... in particular it did not make a new cache for you even in this last build as it said:
```
Skipping cache generation, cache already exists for key: doc-pull/13917
Found one created at 2021-04-22 08:29:13 +0000 UTC
```

From the CircleCI docs, one option is to change the string (for example add `-1` to the end) to force a cache invalidation, push the change, wait for it to finish, then manually restart CircleCI once it comes back green. That second run will tell you if things are cached and restored properly.

**Comment by tupui** [comment]
Ah yes, the cache was not invalidated because it's just based for now on the branch name!  I will try this, thanks 🙇‍♂️

**Comment by mdhaber** [comment]
Would this reduce the time to see documentation artifacts? That would be great since I've always had trouble getting docs to build locally.

**Comment by mdhaber** [comment]
That would be great! Did your last commit not fix the problem?
Not something that I can dive into right now, but sounds really valuable.

File: .circleci/config.yml [hunk 3]
```diff
@@ -12,7 +12,7 @@ _debian: &debian
   name: install Debian dependencies
   command: |
     sudo apt-get update
-    sudo apt-get install libatlas-base-dev liblapack-dev gfortran libgmp-dev libmpfr-dev libfreetype6-dev libpng-dev zlib1g zlib1g-dev texlive-fonts-recommended texlive-latex-recommended texlive-latex-extra texlive-plain-generic latexmk texlive-xetex fonts-freefont-otf
+    sudo apt-get install libopenblas-dev libatlas-base-dev liblapack-dev gfortran libgmp-dev libmpfr-dev libfreetype6-dev libpng-dev zlib1g zlib1g-dev texlive-fonts-recommended texlive-latex-recommended texlive-latex-extra texlive-plain-generic latexmk texlive-xetex fonts-freefont-otf
```
Comment (rgommers) [inline .circleci/config.yml:15]: If you install openblas, you no longer need atlas and liblapack.

Comment (tupui) [reply .circleci/config.yml:15]: ok thanks I will fix it. I've seen this in the linux yaml.

**Comment by tupui** [comment]
ok this is driving me crazy. I don't know what is happening. If I ssh to a node on CircleCI right before the doc build, the cached doc/build is present. But when I run the build I still get that nothing is up to date
```
building [html]: targets for 4213 source files that are out of date
updating environment: [config changed ('version')] 4213 added, 0 changed, 0 removed
```

I don't actually know how this check is done, so might just be something like a date change which is altered during the cache restore. If anyone has ideas...

**Comment by tupui** [comment]
I think I found something on matplotlib's repo: https://github.com/matplotlib/matplotlib/blob/main/.circleci/config.yml

**Comment by tupui** [comment]
Ok so there are 2 issues:

1. Still no caching for the documentation.
   
   There is an improvement, the files are there and no more things like `updating environment: [new config]` or `version`. I am fixing the version in the config file so it's static, otherwise it triggers a full rebuild. But it still shows that all files have changed (no added, so this is already better). I am not sure how Sphinx is comparing, maybe just a timestamp that would need fixing. Maybe something like `export SOURCE_DATE_EPOCH="$(git log -1 --format=%at)"` could help.
2. ASV is taking forever. Now I tried to put the same deps as with the linux job.

Once ASV is sorted out, even without the full dock caching this already adds some value and we could go with that.

I went ahead with #15635 (both to see how it goes and also because I was already working on this CI). So we need to wait on the mailing list.

**Comment by tupui** [comment]
@larsoner can you have another look? At this point I guess point 2 is more important.

**Comment by Smit-create** [comment]
> 2\. ASV is taking forever.

I'm not sure if ASV uses meson build at present. Because `asv.conf.json` uses default build/install commands which are directed to `python setup.py build`, See: https://asv.readthedocs.io/en/stable/asv.conf.json.html#build-command-install-command-uninstall-command.

**Comment by tupui** [comment]
> > 2. ASV is taking forever.
> 
> I'm not sure if ASV uses meson build at present. Because `asv.conf.json` uses default build/install commands which are directed to `python setup.py build`, See: [asv.readthedocs.io/en/stable/asv.conf.json.html#build-command-install-command-uninstall-command](https://asv.readthedocs.io/en/stable/asv.conf.json.html#build-command-install-command-uninstall-command).

I think it does use Meson because there is no rebuild happening in the [log](https://app.circleci.com/pipelines/github/scipy/scipy/11584/workflows/b5506b23-6473-450e-993d-c34e746c4e35/jobs/43620). Or we don't see anything because it's doing it in the background?

**Comment by rgommers** [comment]
`--bench` uses Meson, `--bench-compare` does not. Please just ignore that for now, we shouldn't change that until we've merged `pyproject.toml` updates so Pip uses Meson by default.

**Comment by ilayn** [comment]
> Now we just need to wait for the mailing list.

I think the consensus is established in favor of removal both in the mailing list and in the linked issue. You can do it here or we can do it in another PR

**Comment by tupui** [comment]
> > Now we just need to wait for the mailing list.
> 
> I think the consensus is established in favor of removal both in the mailing list and in the linked issue. You can do it here or we can do it in another PR

Great, then I propose we do this here (just need to figure out the benchmark thing).

**Comment by tupui** [comment]
@ilayn do you have an idea for why the benchmark are taking longer? I am trying now to revert back to
```
pip install --install-option="--no-cython-compile" cython
pip install numpy==1.21.5
```
instead of having latest NumPy and use the wheel for Cython. Here we also have gmpy2 which was not there before (but we should right?). Also we have a difference with `libopenblas-dev`, before we had atlas+lapack.

**Comment by ilayn** [comment]
If it is compiling correctly the benchmark shouldn't matter since it is the same code. But I have no idea. I am not aware of any case needed for gympy2 shouldn't we be using mpmath?

**Comment by rgommers** [comment]
No `gmpy2` is needed, it yields major performance improvements for `mpmath`.

**Comment by tupui** [comment]
> If it is compiling correctly the benchmark shouldn't matter since it is the same code. But I have no idea. I am not aware of any case needed for gympy2 shouldn't we be using mpmath and then it does its own magic?

As far as I understand gmpy2 makes mpmath run faster. So it should even be faster here 😅

**Comment by tupui** [comment]
I think I need to remove the `--dry-run` option from `dev.py` to see all results of the benchmark and maybe see something... @rgommers is there something else I can try before relaunching?

**Comment by rgommers** [comment]
> I think I need to remove the `--dry-run` option from `dev.py` to see all results of the benchmark and maybe see something... @rgommers is there something else I can try before relaunching?

Not sure, this is starting to be hard to follow. There's too much going on here, and I think the only thing you tried to change initially is doc build time, correct? For the rest you seem to be mostly fighting issues we don't have in `main`.

On main we currently have:

image

I would recommend:
- One new PR to remove pdf doc build, and the few minor doc tweaks you are adding here
- One PR to switch the build job to Meson and introduce ccache
- If there's doc build speed-up experimentation to do, keep it here. But I'm not sure if that is worth it, I'd certainly do that locally first. Sphinx caching is poorly implemented and (I believe) timestamp based, so I'd expect it to just not work in CI.

**Comment by tupui** [comment]
> Not sure, this is starting to be hard to follow. There's too much going on here, and I think the only thing you tried to change initially is doc build time, correct? For the rest you seem to be mostly fighting issues we don't have in `main`.
> Sphinx caching is poorly implemented and (I believe) timestamp based, so I'd expect it to just not work in CI.

Yes, the original intent was to cache Sphinx. In the process of trying to fix it, I tried other things.

> I would recommend:
> 
> * One new PR to remove pdf doc build, and the few minor doc tweaks you are adding here
> * One PR to switch the build job to Meson and introduce ccache
> * If there's doc build speed-up experimentation to do, keep it here. But I'm not sure if that is worth it, I'd certainly do that locally first.

Ok, make sense. I will make 2 separate PRs.

**Comment by tupui** [comment]
Ok I have a green CI on gh-15649 where I removed the PDF/latex. I propose to wait for this to be in and then I can make the second PR about Meson.

**Comment by tupui** [comment]
Second PR about Meson is here #15686. I tested a few things and still have issues.

**Comment by tupui** [comment]
This can be closed as the whole workflow has been migrated now.
